### PR TITLE
Refactor F.split_axis

### PR DESCRIPTION
**Pull request by niboshi**
In `F.split_axis`, currently `indices_or_sections` argument is tested whether or not it's `collections.Iterable` (which means only `__iter__` is guaranteed), but actually uses `__len__` and `__getitem__`.

I refactored the logic so that

* Refactored `indices_or_sections` check
  * Check whether or not it's `collections.Sequence` instead (`numpy.ndarray`, which is not a sequence, is specially accepted)
  * Factored out check and normalization code as a method
* Divided`self.indices_or_sections` into separate attributes `self.indices` and `self.sections`, so that no repeated type check is needed.

TODO: revisit after #4281

### CODE REVIEW

**Comment by okuta** [comment]
#4281 is merged.

**Comment by niboshi** [comment]
Rebased.

File: chainer/functions/array/split_axis.py [hunk 1]
```diff
@@ -14,31 +15,67 @@ class SplitAxis(function_node.FunctionNode):
...
-                isinstance(indices_or_sections, collections.Iterable)):
-            for p, n in six.moves.zip(
-                    indices_or_sections, indices_or_sections[1:]):
+        indices, sections = self._get_indices_or_sections(indices_or_sections)
+        assert (indices is None) != (sections is None)
+        self.indices = indices
+        self.sections = sections
+        self.axis = axis
+
+    def _get_indices_or_sections(self, indices_or_sections):
```
Comment (okuta) [inline chainer/functions/array/split_axis.py]: `self` is not used.
How about using module level function?

Comment (niboshi) [reply chainer/functions/array/split_axis.py]: Fixed.

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
Please check example test result.
```
Traceback (most recent call last):
  File "/home/user/.local/lib/python2.7/site-packages/chainer/training/trainer.py", line 309, in run
    entry.extension(self)
  File "examples/seq2seq/seq2seq.py", line 302, in translate
    result = model.translate([model.xp.array(source)])[0]
  File "examples/seq2seq/seq2seq.py", line 77, in translate
    exs = sequence_embed(self.embed_x, xs)
  File "examples/seq2seq/seq2seq.py", line 26, in sequence_embed
    exs = F.split_axis(ex, x_section, 0)
  File "/home/user/.local/lib/python2.7/site-packages/chainer/functions/array/split_axis.py", line 184, in split_axis
    res = SplitAxis(indices_or_sections, axis).apply((x,))
  File "/home/user/.local/lib/python2.7/site-packages/chainer/functions/array/split_axis.py", line 62, in __init__
    indices, sections = _get_indices_or_sections(indices_or_sections)
  File "/home/user/.local/lib/python2.7/site-packages/chainer/functions/array/split_axis.py", line 30, in _get_indices_or_sections
    raise TypeError('indices_or_sections must be integers')
```

**Comment by niboshi** [comment]
Fixed.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**